certbot/auth_handler.py

```diff
@@ -47,242 +41,142 @@ def __init__(self, auth, acme_client, account, pref_challs):
 
-                # Send all Responses - this modifies achalls
-                self._respond(aauthzrs, resp, best_effort)
-
-        # Just make sure all decisions are complete.
-        self.verify_authzr_complete(aauthzrs)
+            # Inform the ACME CA server that challenges are available for validation.
+            for achall, resp in zip(achalls, resps):
+                self.acme.answer_challenge(achall.challb, resp)
 
-        # Only return valid authorizations
-        ret_val = [aauthzr.authzr for aauthzr in aauthzrs
-                   if aauthzr.authzr.body.status == messages.STATUS_VALID]
+            # Wait for authorizations to be checked.
+            self._poll_authorizations(authzrs, max_retries, best_effort)
 
-        if not ret_val:
-            raise errors.AuthorizationError(
-                "Challenges failed for all domains")
+            # Keep validated authorizations only. If there is none, no certificate can be issued.
+            authzrs_validated = [authzr for authzr in authzrs
+                                 if authzr.body.status == messages.STATUS_VALID]
+            if not authzrs_validated:
+                raise errors.AuthorizationError('All challenges have failed.')
 
-        return ret_val
+            return authzrs_validated
 
-    def _choose_challenges(self, aauthzrs):
+    def _poll_authorizations(self, authzrs, max_retries, best_effort):
+        """
+        Poll the ACME CA server, to wait for confirmation that authorizations have their challenges
+        all verified. The poll may occur several times, until all authorizations are checked
+        (valid or invalid), or after a maximum of retries.
+        """
+        authzrs_to_check = {index: (authzr, None)
+                            for index, authzr in enumerate(authzrs)}
+        # Give an initial second to the ACME CA server to check the authorizations
+        sleep_seconds = 1
+        for _ in range(max_retries):
+            # Wait for appropriate time (from Retry-After, initial wait, or no wait)
+            if sleep_seconds > 0:
+                time.sleep(sleep_seconds)
+            # Poll all updated authorizations.
+            authzrs_to_check = {index: self.acme.poll(authzr) for index, (authzr, _)
+                                in authzrs_to_check.items()}
+            # Update the original list of authzr with the updated authzrs from server.
+            for index, (authzr, _) in authzrs_to_check.items():
+                authzrs[index] = authzr
+
+            # Handle failed authorizations: with best effort this is only a warning,
+            # otherwise an exception.
+            authzrs_failed = [authzr for authzr, _ in authzrs_to_check.values()
+                              if authzr.body.status == messages.STATUS_INVALID]
+            if authzrs_failed:
+                # Display report to the user
+                _report_failed_authzrs(authzrs_failed, self.account.key)
```

Comment:
nit: If you look at the output generated by this function, it creates one formatted message with a header and a footer like:
```
 - The following errors were reported by the server:

   Domain: example.org
   Type:   unauthorized
   Detail: Invalid response from
   http://example.org/.well-known/acme-challenge/y3jq7cdUngIddjenKtN-6_613t1YUrvceS-WNCRBJRM
   [2606:2800:220:1:248:1893:25c8:1946]: "<!doctype
   html>\n<html>\n<head>\n    <title>Example Domain</title>\n\n
   <meta charset=\"utf-8\" />\n    <meta http-equiv=\"Content-type"

   Domain: www.example.org
   Type:   unauthorized
   Detail: Invalid response from
   http://www.example.org/.well-known/acme-challenge/Dtch-6y2bYGI_T4Wir-N0H5Oi08ySaYLzVB587CGn1c
   [2606:2800:220:1:248:1893:25c8:1946]: "<!doctype
   html>\n<html>\n<head>\n    <title>Example Domain</title>\n\n
   <meta charset=\"utf-8\" />\n    <meta http-equiv=\"Content-type"

   To fix these errors, please make sure that your domain name was
   entered correctly and the DNS A/AAAA record(s) for that domain
   contain(s) the right IP address.
```
In the `best_effort` case, it's possible we'll call this method multiple times leading to rather unnecessarily verbose and cluttered output.

While this is a fairly minor problem, I think we can fairly easily fix this. What do you think?

Reply:
Yes, it is a good idea. I accumulate now all failed authzrs in case of best_effort, and send one unique report when leaving the loop. I move back to a `break` instead of a `return` for leaving the loop, in order to let the report be displayed even if we reached the max polling retry.

Reply:
Another change that was made as part of this was now when `--allow-subset-of-names` is not used, we still continue to poll all authorizations even if we have failed some. I think this is a good change as I'd personally be slightly annoyed if I failed some challenges, fixed up my config for those domains, and then was told about additional failed challenges for other domains.